core/trino-main/src/main/java/io/trino/type/LikePattern.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.type;
+
+import io.trino.likematcher.LikeMatcher;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public record LikePattern(String pattern, Optional<Character> escape, LikeMatcher likeMatcher)
```

Comment:
Add a Javadoc saying something like:

> Equality for this class is dependent on the pattern and escape alone, as the matcher is expected to be derived from those.

Reply:
Having the matcher provided as an argument is error prone, as it allows for someone to pass a matcher that's not associated with the pattern.

I would turn this into a class (instead of a record) and add a static factory method that instantiates the matcher inside and passes it to a private constructor.